File: src/utils/receiptListParser.ts
/**
 * Parses itemized receipt text (e.g. from HomeGoods / TJ Maxx receipts)
 * into structured item data ready for creation.
 *
 * Expected line format:
 *   DEPT - DESCRIPTION SKU $PRICE T
 *   e.g. "53 - ACCENT FURNISH 252972 $129.99 T"
 */

export type ParsedReceiptItem = {
  name: string;
  sku: string;
  priceCents: number;
};

export type ParseReceiptResult = {
  items: ParsedReceiptItem[];
  skippedLines: string[];
};

/**
 * Regex breakdown:
 *   ^\d+\s*-\s*       department number + dash (discarded)
 *   (.+?)             description (non-greedy)
 *   \s+(\d{4,})       SKU (4+ digit number)
 *   \s+\$?([\d,]+\.\d{2})  price with optional $ and optional commas
 *   \s*T?\s*$          optional taxable flag
 */
const LINE_RE = /^\d+\s*-\s*(.+?)\s+(\d{4,})\s+\$?([\d,]+\.\d{2})\s*T?\s*$/;

function priceToCents(priceStr: string): number {
  const cleaned = priceStr.replace(/,/g, '');
  return Math.round(Number.parseFloat(cleaned) * 100);
}

export function parseReceiptList(text: string): ParseReceiptResult {
  const items: ParsedReceiptItem[] = [];
  const skippedLines: string[] = [];

  const lines = text.split('\n');

  for (const raw of lines) {
    const line = raw.trim();
    if (line.length === 0) continue;

    const match = line.match(LINE_RE);
    if (match) {
      items.push({
        name: match[1].trim(),
        sku: match[2],
        priceCents: priceToCents(match[3]),
      });
    } else {
      skippedLines.push(line);
    }
  }

  return { items, skippedLines };
}

File: LedgeriOS/LedgeriOS/Logic/ReceiptListParser.swift
import Foundation

/// Parses itemized receipt text (e.g. from HomeGoods / TJ Maxx receipts)
/// into structured item data ready for creation.
///
/// Expected line format:
///   DEPT - DESCRIPTION SKU $PRICE T
///   e.g. "53 - ACCENT FURNISH 252972 $129.99 T"
///
/// Ported from RN `receiptListParser.ts`.
enum ReceiptListParser {

    struct ParsedItem: Identifiable, Equatable {
        let id = UUID()
        let name: String
        let sku: String
        let priceCents: Int
    }

    struct ParseResult: Equatable {
        let items: [ParsedItem]
        let skippedLines: [String]

        static func == (lhs: ParseResult, rhs: ParseResult) -> Bool {
            lhs.items.map(\.name) == rhs.items.map(\.name)
            && lhs.items.map(\.priceCents) == rhs.items.map(\.priceCents)
            && lhs.skippedLines == rhs.skippedLines
        }
    }

    /// Regex: dept-dash-description-sku-price-optional-tax-flag
    /// `^\d+\s*-\s*(.+?)\s+(\d{4,})\s+\$?([\d,]+\.\d{2})\s*T?\s*$`
    private static nonisolated(unsafe) let linePattern = /^\d+\s*-\s*(.+?)\s+(\d{4,})\s+\$?([\d,]+\.\d{2})\s*T?\s*$/

    static func parseReceiptText(_ text: String) -> ParseResult {
        var items: [ParsedItem] = []
        var skippedLines: [String] = []

        let lines = text.components(separatedBy: "\n")

        for raw in lines {
            let line = raw.trimmingCharacters(in: .whitespaces)
            guard !line.isEmpty else { continue }

            if let match = line.wholeMatch(of: linePattern) {
                let name = String(match.1).trimmingCharacters(in: .whitespaces)
                let sku = String(match.2)
                let priceStr = String(match.3).replacingOccurrences(of: ",", with: "")
                let priceCents = Int((Double(priceStr) ?? 0) * 100)

                items.append(ParsedItem(name: name, sku: sku, priceCents: priceCents))
            } else {
                skippedLines.append(line)
            }
        }

        return ParseResult(items: items, skippedLines: skippedLines)
    }
}
